src/cpu/simple/probes/looppoint_analysis.hh

```diff
@@ -0,0 +1,331 @@
+     * the localBBV.
+     */
+    void updateLocalBBV(const Addr pc);
+
+  public:
+    std::unordered_map<Addr, uint64_t> getLocalBBV() const
+    {
+        return localBBV;
+    };
+
+    void clearLocalBBV()
+    {
+        localBBV.clear();
+    };
+};
+
+class LooppointAnalysisManager: public SimObject
+{
+  public:
+    LooppointAnalysisManager(const LooppointAnalysisManagerParams &params);
+
+    /**
+     * This function is called by the LooppointAnalysis probe listener when it
+     * finds a valid backward branch that can be used as a marker.
+     * Specifically, it increments the backwardBranchCounter for the input PC.
+     */
+    void countBackwardBranch(const Addr pc);
+
+    /**
+     * This function is called by the LooppointAnalysis probe listener when it
+     * reaches the branch instruction for a valid basic block.
+     * Specifically, it increments the globalBBV for the input PC.
+     */
+    void updateGlobalBBV(const Addr pc);
+
+  private:
+    /**
+     * This counter is for the valid backward branches that we consider as
+     * candidates for marking the execution points.
+     * The key is the Program Counter address of the backward branch, and the
+     * value is the number of times the backward branch is executed.
+     */
+    std::unordered_map<Addr, uint64_t> backwardBranchCounter;
+
+    /**
+     * This is the global basic block vector that contains the count of each
+     * basic block that is executed.
+     * The key is the Program Counter address of the basic block's branch
+     * instruction, and the value is the number of times the basic block is
+     * executed.
+     */
+    std::unordered_map<Addr, uint64_t> globalBBV;
+
+    /**
+     * This map stores the number of instructions in each basic block.
+     * The key is the Program Counter address of the basic block's branch
+     * instruction, and the value is the number of instructions in the basic
+     * block.
+     */
+    std::unordered_map<Addr, uint64_t> bbInstMap;
+
+    /**
+     * This variable stores the number of instructions that we used to define
+     * a region. For example, if the regionLength is 100, then every time when
+     * there are 100 valid instructions executed globally through all the
+     * cores, we will consider that as a region.
+     */
+    uint64_t regionLength;
+
+    /**
+     * This is a counter for the globally executed instructions. We use this to
+     * compare with the regionLength to determine if we reach the end of a
+     * region.
+     */
+    uint64_t globalInstCounter;
+
+    /**
+     * This variable stores the Program Counter address of the most recent
+     * valid backward branch that we consider as a marker.
+     */
+    Addr mostRecentBackwardBranchPC;
+
+    /**
+     * This set stores the Program Counter addresses of the valid backward
+     * branches.
+     */
+    std::unordered_set<Addr> backwardBranchPC;
```

Comment:
Is there a reason you can't use `backwardBranchCounter` for this?

Reply:
It is a great catch! I removed `backwardBranchPC`and started using `backwardBranchCounter` instead now.